### DISCUSSION TITLE
Implementation of strict types for populate array (dot notation and string literal types)

### DISCUSSION THREAD

**Discussion by Char2sGu** [mikro-orm/mikro-orm#2054] [Ideas]
>  Strict typings for populate array (dot notation and string literal types)

This is a feature in the [task list of v5](https://github.com/mikro-orm/mikro-orm/issues/1623#issue-844206844) which is not yet implemented, I noticed it just recently and get really excited because I have already implemented it a long time ago in [nest-mikro-crud](https://github.com/TheNightmareX/nest-mikro-crud)! You can check out the type implementation at [src/types/relation-path.type.ts](https://github.com/TheNightmareX/nest-mikro-crud/blob/main/src/types/relation-path.type.ts) and [src/utils/extract-path.type.ts](https://github.com/TheNightmareX/nest-mikro-crud/blob/main/src/utils/extract-path.type.ts)!

You can try it by yourself:

```ts
import { Collection } from '@mikro-orm/core';

export type ValueOf<T> = T[keyof T];

export type ItemTypeIfIterable<T> = T extends Iterable<infer R> ? R : T;

/**
 * Extract all the keys that meet the condition in the target and its sub-objects and
 * connect the keys using the separator.
 *
 * @param Visited - **DO NOT** specify this type!!! Used to prevent circular references.
 *
 * @example
 * type ObjType = {
 *   str: string;
 *   num: number;
 *   obj: { str: string; num: number; circular: ObjType };
 *   arr: { str: string; num: number; circular: ObjType }[];
 * };
 *
 * type T = ExtractPath<ObjType, string | number, number, "/">;
 * // type T = "str" | "obj/str" | "arr/str"
 */
export type ExtractPath<
  Target,
  Condition,
  Exclusion = never,
  Separator extends string = '.',
  Visited = Target
> = ValueOf<
  {
    [K in string & keyof Target]:
      | (ItemTypeIfIterable<Target[K]> extends Exclusion
          ? never
          : ItemTypeIfIterable<Target[K]> extends Condition
          ? K
          : never)
      | (ItemTypeIfIterable<Target[K]> extends Visited
          ? never
          : // @ts-expect-error - this IS NOT an infinite loop!!!!
            `${K}${Separator}${ExtractPath<
              ItemTypeIfIterable<Target[K]>,
              Condition,
              Exclusion,
              Separator,
              Visited | ItemTypeIfIterable<Target[K]>
            >}`);
  }
>;

export type RelationPath<Entity> = ExtractPath<
  Entity,
  null | Record<string, any>,
  Function | Date
>;

interface User {
  id: number;
  classroomsCreated: Collection<Classroom>;
  joinApplications: Collection<JoinApplication>;
  memberships: Collection<Membership>;
  createdAt: Date;
  updatedAt: Date;
}

interface Membership {
  id: number;
  owner: User;
  classroom: Classroom;
  createdAt: Date;
  updatedAt: Date;
}

interface Classroom {
  id: number;
  creator: User;
  joinApplications: Collection<JoinApplication>;
  memberships: Collection<Membership>;
  createdAt: Date;
  updatedAt: Date;
}

interface JoinApplication {
  id: number;
  owner: User;
  classroom: Classroom;
  createdAt: Date;
  updatedAt: Date;
}

type T = RelationPath<Membership>;

const test: T = 
// classroom
// classroom.creator
// classroom.creator.classroomsCreated
// classroom.creator.joinApplications
// classroom.creator.memberships
// classroom.joinApplications
// classroom.memberships
// owner
// owner.classroomsCreated
// owner.classroomsCreated.creator
// owner.classroomsCreated.joinApplications
// owner.classroomsCreated.memberships
// owner.joinApplications
// owner.memberships

```

There should be a TypeScript Playground link, but the the playground seems to be broken now, I will add the link once the playground is fixed.

I really want to implement it and create a pull request by myself, but because I am still a high school student and am quite new in open source projects, I don't have enough time to achieve that by myself and also afraid of messing up, so I hope someone could adopt my implementation in nest-mikro-orm and implement this awesome feature in MikroORM v5. Btw, if possible, I hope to have a line of comment in the relevant code like `// implementation from nest-mikro-orm by TheNightmareX`, I really need to prove myself, thanks!

**Comment by B4nan** [comment]
Hi there, unfortunately you are comming a bit too late, as noted in #1623 I already have a working implementation of this. 

If you want to contribute, I just raised another question on stack overflow about a follow up - the `Loaded` return type mapping:

https://stackoverflow.com/questions/68454737/mapping-of-strict-path-notation-array-type

Will review the code, maybe it can help, but the main blocker right now is the `Loaded` type and in general the integration to current code base. 

I am not against having a comment in the code to point to the source, but much better would be if you could actually help integrating it into codebase and send a PR instead. The comment would be maybe nice for you, but would have pretty much no value for the project or its contributors. If you want to prove yourself, you should strive to have a commit in the master branch, not a comment in the code :] 

And I hope you understand that when I already have a working implementation I wrote myself and therefore fully understand it, it is better to use that instead of reinventing the wheel. 

Btw I did check the code quickly, and it fails to handle recursion (e.g. `User` have property of type `User`), so maybe you might wanna check my implementation which does not suffer from that. I also see it won't work for more than 2 levels, another things that works in my version.

**Reply by Char2sGu** [reply]
Okay...  I'm too hurried and didn't see that you already have a better implementation... :-(

I will help developing this awesome ORM as much as possible, it really helped me a lot!
